### DISCUSSION TITLE
Upgrade to latest release of this repo?

### DISCUSSION THREAD

**Discussion by cap10morgan** [Joshua-Riek/ubuntu-rockchip#81] [Q&A] [answered]
Is it possible to upgrade an existing system to the latest release of this repo (i.e. _not_ Ubuntu itself)? Will `apt update && apt upgrade -y` get you everything? Or is it more of a "backup and re-flash" kind of thing currently?

Thanks for this it's amazing!

**Comment by Joshua-Riek** [comment, marked as answer]
Sadly I don't have any sort of upgrade process yet. I plan to create a few scripts to update the kernel, firmware, and bootloader. But, this would not account for system changes such as modifications to the default hostname or partition information. Also, I would like to make a script to update some of the GPU-related packages, I have been hesitant about this because everything is still experimental and an update could break things.

**Reply by cap10morgan** [reply]
OK backup and re-flash it is! Thanks!

**Reply by cap10morgan** [reply]
Oh, another thought I had: one potential incremental approach could be to put `/home` on a different partition in the image and then offer an update tool that replaces everything else. That would still reset all of the things you mention like hostname, but would be an improvement. Users could put scripts in their home dir that set these things the way they want after an update, for example.

**Reply by Joshua-Riek** [reply]
That could work. I was not expecting this many people to be interested in my project here, and updates were not initially on my mind. But I'm learning quite a lot from this and keep coming across many improvements I want to release to everyone. For example, I found out that I can add the Ubuntu OEM installer GUI so everyone can set their language, timezone, and configure user account information. In short, I have a lot to work on for this project.

**Reply by cap10morgan** [reply]
These improvements in 1.11 are awesome! 👍🏻 🎉

**Comment by ferrellsl** [comment]
Although I've never done this, setting up a repository server appears to be pretty mundane according to this site:

https://www.linuxtechi.com/setup-local-apt-repository-server-ubuntu/

You'd need another Pi 5 running Ubuntu server with a routable, static IP address and instead of mirroring archive.ubuntu.com you'd mirror your own packages.

**Reply by Joshua-Riek** [reply]
I have this setup on my home server to mirror ports.ubuntu.com. It helps speed up apt downloads for my ubuntu builds. I'll use something similar for firmware, kernel, and bootloader updates.

I have been considering removing the GPU-related packages from my GitHub repo and using the PPA from https://launchpad.net/~liujianfeng1994, so it's easy to update with apt, and I don't have to manage it. However, I would lose version control of these packages. Do you have any thoughts on this, or think it might be a good idea?

**Reply by ferrellsl** [reply]
I think that's a good idea. But I'm C/C++ programmer, so I'm probably not the best person to advise you on Linux system administration. 
And you could always revert back to your previous method of updating your distro if managing your own repo server becomes too much of a hassle or version control becomes a hindrance.

**Reply by fvillanustre** [reply]
Well, you can always manage version control through either a shell script to install specific versions from the PPA from liujianfeng1994 or dependency management (requires, etc.). That would certainly be preferable and could be done during the first run.

**Comment by Joshua-Riek** [comment]
Today, I have been playing with Launchpad and created two mirrors from liujianfeng1994's PPAs. This should allow me to distribute package updates and test changes made to liujianfeng1994's PPA before updating my mirror and distributing those changes to users.

https://launchpad.net/~jjriek/+archive/ubuntu/rockchip-multimedia
https://launchpad.net/~jjriek/+archive/ubuntu/panfork-mesa

**Reply by ferrellsl** [reply]
Sorry if the answer is obvious but I'm still somewhat of a Linux novice.  With the release of v1.11, has v1.11 already been pre-configured to receive updates/upgrades from your mirror? Or do we have to add the PPA, certs, etc, before we start receiving updates from your mirror?

**Reply by Joshua-Riek** [reply]
It's included, by default, no need to do any configuration! For now, it's only set up for GPU / multimedia package updates.

I still need to add the kernel, bootloader, and firmware. However, I now have a basic process in place to distribute updates. I don't know if I can add them to v1.11 or if a v1.12 will be required as I will still need to make several changes.

**Reply by ferrellsl** [reply]
Thanks so much.  Amazing work in such a short time.

**Comment by Joshua-Riek** [comment]
I wanted to share some more progress I have made. Launchpad has a learning curve, but I finally have a packaged kernel uploaded. However, I do not expect this to be in the next v1.12 release I still need to come up with a proper versioning scheme and update process on my end.

https://launchpad.net/~jjriek/+archive/ubuntu/orange-pi5
https://launchpad.net/~jjriek/+archive/ubuntu/orange-pi5-staging
